### PR TITLE
Add custom domain support to callable functions

### PR DESCRIPTION
**Pull request by samtstern**
Hey there! So you want to contribute to a Firebase SDK? 
Before you file this pull request, please read these guidelines:

### Discussion

[Googlers: see API review here](https://docs.google.com/document/d/1cSIcGCW2QR1OLHX4RDCVNhpBUolnG6tfNpJBR2pn1aM/edit#)

### Testing

  * Added new unit tests

### API Changes

  * API reviewed

### CODE REVIEW

**Review by hsubox76 — COMMENTED** [review]
Looks good, one question, we have to update packages-exp/functions-exp as well with any changes. Would you like to add them here, in a separate PR, or do you want me to take care of it? (I don't mind, should be simple enough to mirror the changes over.)

**Comment by samtstern** [comment]
@hsubox76 I'll give it a try!

File: packages-exp/functions-exp/src/service.ts [hunk 1]
```diff
@@ -94,6 +96,16 @@ export class FunctionsService implements _FirebaseService {
         return Promise.resolve(resolve());
       };
     });
+
+    // Resolve the region or custom domain overload by attempting to parse it.
+    try {
+      const url = new URL(regionOrCustomDomain);
+      this.customDomain = url.origin;
+      this.region = 'us-central1';
```
[the remark this answers is not in the document]
Comment (samtstern) [reply packages-exp/functions-exp/src/service.ts:104]: Wow yeah that was a big mistake on my part, thanks!

File: packages-exp/functions-exp/src/service.ts [hunk 2]
```diff
@@ -84,7 +86,7 @@ export class FunctionsService implements _FirebaseService {
     readonly app: FirebaseApp,
     authProvider: Provider<FirebaseAuthInternalName>,
     messagingProvider: Provider<FirebaseMessagingName>,
-    readonly region: string = DEFAULT_REGION,
+    regionOrCustomDomain: string = DEFAULT_REGION,
```
[the remark this answers is not in the document]
Comment (samtstern) [reply packages-exp/functions-exp/src/service.ts:89]: So if I make it `private` then it remains a private instance field, but I want it to just be a constructor arg that's not saved outside this scope.  So I think it's correct as-is?

**Review by hsubox76 — APPROVED** [review]
LG, by the way 
1) I notice in the design doc the signature is given as `firebase.functions(region)` but it should be `firebase.functions(app, region)`
2) But it looks like that is not documented and in fact there is currently a bug where setting region doesn't work through `firebase.functions(app, region)` but only through
    ```
    const app = firebase.initializeApp({...});
    app.functions(region);
    ```
    So we will fix that separately.

**Comment by samtstern** [comment]
@hsubox76 says I can ignore the test.

**Comment by LanderBeeuwsaert** [comment]
Hello @samtstern, a question we have about this PR:
is this related to https://firebase.google.com/docs/hosting/functions#create_an_http_function_to_your_hosting_site ?
Because we actually wanted to use this possibility in our application to make it easier for clients to interact with our functions, however it is until now only possible to use this in the US region.
Will this change make it possible to use in functions deployed in Europe as well?
Or am I totally off topic here?

**Comment by samtstern** [comment]
@LanderBeeuwsaert no this won't really change that.  You could already use callable functions in your app from any region.  This is about allowing you to use [functions redirects](https://firebase.google.com/docs/hosting/full-config#rewrite-functions) with Callable Functions.

So if your callable function (still in `us-central1`) is called `fooBar` you could call it from `https://mydomain.com/fooBar`.  The main advantage of this is that it reduces lock-in.  If one day you want to leave Cloud Functions and move to say, Cloud Run or Google Compute Engine, your client side code would not have to know about it.

**Comment by LanderBeeuwsaert** [comment]
@samtstern What I seem to understand out of your response is that this PR enables the possibility for functions redirects , but only still for us-central1. Correct?
Are there any plans to enable this sort of redirect for functions deployed in other regions (e.g. europe-west3)?

**Comment by samtstern** [comment]
@LanderBeeuwsaert we'd like to enable redirects to other regions one day but that won't require any changes in the JS SDK!  That's a Hosting backend change.

**Comment by LanderBeeuwsaert** [comment]
top, thanks for the quick response!

**Pull request merged**